feat: cli installer

erunion opened the pull request:
| 🚥 Fix RM-3414, RM-4686 |
| :-- |

## 🧰 Changes

This contains all of the groundwork for an `api` CLI installer that will:

* Fetch an API definition from a local or external source.
* Install that API definition into a new `.apis/` directory.
* Configure that API into an `.apis/api.json` file (modeled after Node's `package.json`).
* Generate a TypeScript library for that API and save it into a directory in `.apis/`.
  * It can also generate a JS library + TS types if you supply `--lang=js` or `--lang=javascript` to the installer.
* Use NPM to install required packages in `.apis/<API>` for that codegen'd library to work.
* Use NPM to install a local package reference to this library as `@api/<API>` in your local working directory.

There's still a lot to do here but there's enough here that I should be able to tag an initial v5 beta with.

### `.apis/` directory

This directory, and really the installation process, has been modeled after the way that NPM works:

```
.api/
├── api.json             // The `package-lock.json` equivalent that records everything that's
|                        // installed, when it was installed, what the original source was,
|                        // and what version of `api` was used.
└── apis/
    ├── readme/
    |   ├── node_modules/
    │   ├── index.js     // We may offer the option to export a raw TS file for folks who want
    |   |                // that, but for now it'll be a compiled JS file.
    │   ├── types.ts     // All types for their SDK, ready to use in an IDE.
    │   |── openapi.json
    │   └── package.json
    └── petstore/
        ├── node_modules/
        ├── index.js
        ├── types.ts
        ├── openapi.json
        └── package.json
```

## 🧬 QA & Testing

Run the following code and answer the questions it prompts you with:

```sh
$ npm run build
$ cd packages/api
$ node dist/bin.js install PATH_TO_A_SPEC
```

* [ ] Does the API you exist in `.api/apis`?
* [ ] Is it in `.api/api.json`?
* [ ] Was `@api/YOUR API` installed to your current directory and added into your `package.json` file?
* [ ] Does the codegen'd code in `.api/apis/YOUR API/index.ts` match up with your spec?

If you need a spec to test with you can use https://raw.githubusercontent.com/readmeio/oas-examples/main/3.0/json/petstore-simple.json.

erunion commented on packages/api/src/cli/storage.ts:
```diff
@@ -0,0 +1,297 @@
+import type { OASDocument } from 'oas/@types/rmoas.types';
```
A lot of this is duped from the `src/cache` library that powers the dynamic, non-installed `api`. I'd love to merge the two so if you're using a dynamic `api` you already have your spec installed in `.api` and can easily create an SDK for it, but that'd require us to know the identifier for the spec upfront and we don't have a way to ask these users that without overcomplicating the dynamic `api` API.

erunion commented on packages/api/src/cli/storage.ts:
```diff
@@ -0,0 +1,297 @@
+  }
+
+  static generateIntegrityHash(definition: OASDocument) {
+    return ssri
```
[ssri](https://npm.im/ssri) is the same library that NPM uses to generate integrity hashes in `package-lock.json` files.

erunion commented on packages/api/test/integration.test.ts:
```diff
@@ -143,14 +140,12 @@ describe('integration tests', function () {
         expect(res.requestBody.split(`--${res.boundary}`).filter(Boolean)).to.deep.equal([
           '\r\nContent-Disposition: form-data; name="orderId"\r\n\r\n1234\r\n',
           '\r\nContent-Disposition: form-data; name="userId"\r\n\r\n5678\r\n',
-          isNode18
```
I no longer need all this mess because the changes that broken this stuff in [undici](https://npm.im/undici) were reverted and then published as part of Node 18.4.

erunion replied:
I'm glad undici cleaned their mess up.

kanadgupta reviewed (changes requested):
Opened #460 which describes the issues I'm having loading the package. Aside from that, the CLI flow for installing the package is real smooth 🧈  

haven't looked through all the code yet, but noticed a few things that are called out below!

erunion commented on packages/api/package.json:
```diff
@@ -4,8 +4,12 @@
   "main": "dist/index.js",
   "types": "dist/index.d.ts",
+  "bin": {
+    "rdme": "dist/bin.js"
```
lmao. yeah uh oops

domharrington reviewed (approved):
For the initial question:
> ? What would you like to identify this API as? This will be how you import the SDK. (e.g. entering `petstore` would result in `@api/petstore`)

Could we try and infer a good enough name, maybe by slugifying the `info.title`, or defaulting to the filename that we're fetching?

I ran this command here to test this out:

```sh
node dist/bin.js install https://raw.githubusercontent.com/OAI/OpenAPI-Specification/main/examples/v3.0/petstore.json
```

Everything seemed to work, but `@api/petstore` is missing from `api/packages/api/package.json`. 

---

There's a lot to this code, and I don't fully understand all of it, but it seems good enough to cut a first release for sure.

erunion commented on packages/api/src/bin.ts:
```diff
@@ -0,0 +1,20 @@
+import { Command } from 'commander';
```
They've also fixed a lot of the issues I was having with it when rdme was still on it with subcommands no longer overtaking `--version` arguments.

erunion commented on packages/api/src/cli/codegen/languages/typescript.ts:
```diff
@@ -84,6 +106,35 @@ export default class TSGenerator extends CodeGenerator {
+      main: './index.ts',
+    };
+
+    fs.writeFileSync(path.join(installDir, 'package.json'), JSON.stringify(pkg, null, 2));
```
Yep!

erunion commented on packages/api/src/cli/codegen/languages/typescript.ts:
```diff
@@ -84,6 +106,35 @@ export default class TSGenerator extends CodeGenerator {
+      }
+    });
+
+    return execa('npm', ['install', storage.getIdentifierStorageDir(), '--dry-run']).then(res => {
```
It's supposed to but I'm just now realizing that I left in the `--dry-run` here. No wonder `@api/petstore` hasn't been getting installed. I'll fix this and then add a bunch of comments in this function to clarify what's happening because it definitely isn't clear at the moment.

erunion commented on packages/api/src/cli/codegen/languages/typescript.ts:
```diff
@@ -233,7 +284,16 @@ sdk.server('https://eu.api.example.com/v14');`)
     });
 
-    // const result = project.emitToMemory();
+    if (this.outputJS) {
```
Yep

erunion commented on packages/api/src/cli/storage.ts:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    // Create our main `.api/` directory.
+    if (!fs.existsSync(Storage.dir)) {
```
None that I'm aware of. I'm mostly using `*Sync` in a few spots in this file because the methods they're in are nested within some other promises and it's easier to use the sync version of `fs` then having to deal with nested promises or promises in a foreach.

erunion commented:
> Could we try and infer a good enough name, maybe by slugifying the info.title

@domharrington There's a few problems with this:

* If we ask ReadMe customers to break up their specs to get around our bad upload timeouts they're likely not going to change `info.title` resulting in it being duped across multiple specs. When someone tries to install their 2+ files they'll have problems.
* `info.title` also doesn't have any length limits in the [spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#infoObject). It'll likely be a short string but it could also be a huge Markdown blob.

> or defaulting to the filename that we're fetching?

The installer supports a few different methods of installing a spec:

* `@developers/v2.0#nysezql0wwo236`
  * We can't reliably use `developers` out of the API Registry accessor because thats the project subdomain and if they're installing multiple specs they'll already have `developers` installed.
* https://raw.githubusercontent.com/readmeio/oas-examples/main/3.0/json/petstore-simple.json
  * We could pop `petstore-simple` off this but we also aren't always guaranteed to have a URL like this. Maybe the URL is https://example.com/openapi or https://oas.example.com 
* `./petstore.json`
  * Though it'll likely be as rare as someone installing multiple specs from the same project this might run the risk of someone installing two different specs that happen to share the same filename.

For all these reasons I thought it's easier for now to ask upfront what they want it to be called.

erunion commented:
I pushed up a big beefy in https://github.com/readmeio/api/pull/427/commits/23e675d74a44b1e9649d84e877213cae2e82866b that should address every problem that @kanadgupta had with codegen.

* CLI now prompts you for the language. If you select JS it'll ask if you want a library ready for CommonJS or ESM.
  * Our TS exports are raw source files so we don't need to worry about CJS or ESM nonsense there. If you're on TS it'll just work. <sup><sup>🤞</sup></sup>
* I fixed a whole shitload of issues I found while testing all this and debugging CJS vs ESM issues with `ts-morph`.
* Updated user agents in the generated SDKs to now be contextual to the identiifer the user used for the library + the `info.version` from the spec and the version of `api` that was used. Example: `readme/2.0.0 (api/5.0.0-beta)`.

The `api` path in `package.json` still needs to get changed manually (see https://github.com/readmeio/api/pull/460#discussion_r910431870) when testing this locally but everything else should be in tip top shape now.

@kanadgupta @domharrington if you don't mind giving this another look I'd appreciate it.

kanadgupta reviewed (approved):
Aside from an issue I'm seeing when trying to make an authenticated request (discussing this with @erunion offline), this works beautifully and I think it's ready for beta! Amazing work 🥳

kanadgupta reviewed (approved):
Confirmed that the changes in https://github.com/readmeio/api/pull/427/commits/7c69b2849bb2ab8ca4a5d7f3b0ba92dc401ea89d fixed the auth issues!
